Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs

```diff
@@ -1,110 +1,406 @@
+
+                foreach (var ent in _trackedDevices)
+                    ev.Entities.Add(ent);
+
+                break;
+            }
+        }
+    }
+
+    private void OnUpdateRequestReceived(EntityUid uid, PowerMonitoringConsoleComponent component, RequestPowerMonitoringUpdateMessage args)
+    {
+        UpdateUIState(uid, component, GetEntity(args.FocusDevice), args.Session);
+    }
+
+    private void OnPowerGridCheckStarted(ref GameRuleStartedEvent ev)
+    {
+        if (HasComp<PowerGridCheckRuleComponent>(ev.RuleEntity))
+            _powerNetAbnormalities = true;
+    }
+
+    private void OnPowerGridCheckEnded(ref GameRuleEndedEvent ev)
+    {
+        if (HasComp<PowerGridCheckRuleComponent>(ev.RuleEntity))
+            _powerNetAbnormalities = false;
+    }
+
+    public void UpdateUIState(EntityUid uid, PowerMonitoringConsoleComponent powerMonitoring, EntityUid? focus, ICommonSession session)
+    {
+        if (!_userInterfaceSystem.TryGetUi(uid, PowerMonitoringConsoleUiKey.Key, out var bui))
+            return;
+
+        var consoleXform = Transform(uid);
+
+        if (consoleXform?.GridUid == null)
             return;
-        if (!Resolve(target, ref ncComp))
+
+        var gridUid = consoleXform.GridUid.Value;
+
+        if (!TryComp<MapGridComponent>(gridUid, out var mapGrid))
             return;
 
-        var totalSources = 0.0d;
-        var totalLoads = 0.0d;
-        var sources = new List<PowerMonitoringConsoleEntry>();
-        var loads = new List<PowerMonitoringConsoleEntry>();
-        PowerMonitoringConsoleEntry LoadOrSource(Component comp, double rate, bool isBattery)
+        // The grid must have a NavMapComponent to visualize the map 
+        EnsureComp<NavMapComponent>(gridUid);
+
+        // Data to be send to the client
+        var totalSources = 0d;
+        var totalBatteryUsage = 0d;
+        var totalLoads = 0d;
+        var allEntries = new List<PowerMonitoringConsoleEntry>();
+        var sourcesForFocus = new List<PowerMonitoringConsoleEntry>();
+        var loadsForFocus = new List<PowerMonitoringConsoleEntry>();
+        var cableNetwork = GetPowerCableNetworkBitMask(gridUid, mapGrid);
+        var focusNetwork = new Dictionary<Vector2i, NavMapChunkPowerCables>();
+        var flags = PowerMonitoringFlags.None;
+
+        // Power grid anomalies event is in-progress
+        if (_powerNetAbnormalities)
+            flags |= PowerMonitoringFlags.PowerNetAbnormalities;
+
+        // Power consumers 
+        var powerConsumerQuery = AllEntityQuery<PowerConsumerComponent>();
+        while (powerConsumerQuery.MoveNext(out var ent, out var powerConsumer))
         {
-            var md = MetaData(comp.Owner);
-            var prototype = md.EntityPrototype?.ID ?? "";
-            return new PowerMonitoringConsoleEntry(md.EntityName, prototype, rate, isBattery);
+            if (powerConsumer.ReceivedPower >= RoguePowerConsumerThreshold)
```

Comment:
Does this mean that if there is an active power sink, but it receives less than 100kW input, it won't trigger the warning?

Reply:
That's correct. But even with a single SMES on the same network, it will hit that threshold within a minute